File: services/mirror-worker/src/replay/order-created-started.ts
import axios from 'axios';
import { PoolClient } from 'pg';
import { config } from '../config';
import { log } from '../log';
import { ProdUserLog, prodSelectOne } from '../db/prod';
import {
  insertMapping,
  lookupMapping,
  stagingSelectOne,
  withStagingTx,
} from '../db/staging';
import { getStagingApiToken } from '../runtime';
import {
  translateAreaId,
  translateEquipmentId,
  translateSiteId,
} from '../translate';

// order-created-started is the operator's "create + start in one shot" flow.
// This is the most important replay handler for closing Gap 1 — without it,
// new POs created via the operator's New PO button never appear on staging
// until the manual backfill script runs.
//
// Sample prod payload:
//   {
//     idArea, idSite,
//     idOrder, timestamp,
//     idEquipment, idEnterprise,
//     equipmentSetup: [{id, position}, ...],
//     unitMultiplier,
//     productionOrderQuantity
//   }
//
// Output: POST /api/production-orders/create-and-start (CreateAndStartDto)
//   { idEnterprise, idSite, idArea, idEquipment, idOrder,
//     productionOrderQuantity, timestamp }
//
// We use the prod timestamp directly. This may collide with historical
// finished POs on the same equipment via getOrderDateConflict — but it
// preserves prod-faithful ts_start. The backfill script uses now-5s as
// a workaround for that specific case; for real-time replays of fresh
// prod actions, prod's timestamp is the right value.
interface OrderCreatedStartedPayload {
  idArea: number;
  idSite: number;
  idOrder: number | string;
  timestamp: string;
  idEquipment: number;
  idEnterprise: number;
  productionOrderQuantity: number | string;
}

export async function replayOrderCreatedStarted(
  _client: PoolClient,
  row: ProdUserLog,
): Promise<void> {
  const payload = row.payload as OrderCreatedStartedPayload;
  if (!payload || typeof payload !== 'object') {
    throw new Error('order-created-started payload missing or not an object');
  }
  if (typeof payload.idEquipment !== 'number') {
    throw new Error('order-created-started payload missing idEquipment');
  }
  if (payload.idOrder === undefined || payload.idOrder === null) {
    throw new Error('order-created-started payload missing idOrder');
  }

  const idOrderNum = Number(payload.idOrder);

  const stagingEquipmentId = await translateEquipmentId(payload.idEquipment);
  const stagingSiteId = await translateSiteId(payload.idSite);
  const stagingAreaId = await translateAreaId(payload.idArea);

  const prodPoRow = await prodSelectOne<{ id_production_order: string }>(
    `SELECT id_production_order::text
       FROM production_orders
      WHERE id_enterprise = $1 AND id_order = $2
      ORDER BY id_production_order DESC LIMIT 1`,
    [config.prodEnterpriseId, idOrderNum],
  );
  if (!prodPoRow) {
    throw new Error(
      `no prod production_order found for idEnterprise=${config.prodEnterpriseId} idOrder=${idOrderNum}`,
    );
  }
  const prodPoId = BigInt(prodPoRow.id_production_order);

  const existing = await lookupMapping('production_order', config.sourceName, prodPoId);
  if (existing !== undefined) {
    log.info(
      { sourceLogId: row.id_user_logs, prodPoId: prodPoId.toString(), stagingPoId: existing.toString() },
      'order-created-started already mapped, skipping',
    );
    return;
  }

  const body = {
    idEnterprise: config.stagingEnterpriseId,
    idSite: stagingSiteId,
    idArea: stagingAreaId,
    idEquipment: stagingEquipmentId,
    idOrder: idOrderNum,
    productionOrderQuantity: Number(payload.productionOrderQuantity),
    timestamp: payload.timestamp,
  };

  const res = await axios.post(
    `${config.stagingApi.baseUrl}/api/production-orders/create-and-start`,
    body,
    {
      params: {
        token: getStagingApiToken(),
        idEnterprise: config.stagingEnterpriseId,
      },
      headers: {
        'x-user': row.nm_user ?? 'mirror-worker',
        'X-Mirror-Source': config.sourceName,
      },
      validateStatus: () => true,
      timeout: 15_000,
    },
  );
  if (res.status >= 400) {
    throw new Error(
      `staging /api/production-orders/create-and-start returned ${res.status}: ${JSON.stringify(res.data)}`,
    );
  }

  const created = await stagingSelectOne<{ id_production_order: string }>(
    `SELECT id_production_order::text
       FROM production_orders
      WHERE id_enterprise = $1 AND id_order = $2
      ORDER BY id_production_order DESC LIMIT 1`,
    [config.stagingEnterpriseId, idOrderNum],
  );
  if (!created) {
    throw new Error(
      `create-and-start returned ${res.status} but staging PO with idOrder=${idOrderNum} not found`,
    );
  }
  const stagingPoId = BigInt(created.id_production_order);
  await withStagingTx(async (c) => {
    await insertMapping(c, {
      entityType: 'production_order',
      source: config.sourceName,
      prodId: prodPoId,
      stagingId: stagingPoId,
      sourceLogId: BigInt(row.id_user_logs),
    });
  });

  log.info(
    {
      sourceLogId: row.id_user_logs,
      prodPoId: prodPoId.toString(),
      stagingPoId: stagingPoId.toString(),
      idOrder: idOrderNum,
      status: res.status,
    },
    'replayed order-created-started',
  );
}

File: services/mirror-worker-go/internal/replay/order_created_started.go
package replay

import (
	"context"
	"encoding/json"
	"fmt"
	"log/slog"
	"net/http"

	"github.com/jackc/pgx/v5"
	"github.com/packiot/packiot-stack-alpha/services/mirror-worker-go/internal/config"
	"github.com/packiot/packiot-stack-alpha/services/mirror-worker-go/internal/db"
	"github.com/packiot/packiot-stack-alpha/services/mirror-worker-go/internal/translate"
)

// order-created-started — operator's "create + start in one shot" flow.
// Most important replay handler for closing Gap 1: new POs created via
// the operator's New PO button stay invisible to staging until this fires.
//
// Uses prod's timestamp directly. Faithful but may collide via edge-api's
// getOrderDateConflict against historical finished POs on the same equipment.
// Backfill scripts use a now-5s workaround; real-time replays of fresh prod
// actions don't have that problem (prod's ts_start is current).
type OrderCreatedStartedPayload struct {
	IDArea                  int         `json:"idArea"`
	IDSite                  int         `json:"idSite"`
	IDOrder                 json.Number `json:"idOrder"`
	Timestamp               string      `json:"timestamp"`
	IDEquipment             int         `json:"idEquipment"`
	ProductionOrderQuantity json.Number `json:"productionOrderQuantity"`
}

func OrderCreatedStarted(
	cfg *config.Config,
	t *translate.Translator,
	prodDB *db.Prod,
	stagingDB *db.Staging,
	apiToken func() string,
	httpc *http.Client,
	logger *slog.Logger,
) Handler {
	return func(ctx context.Context, tx pgx.Tx, row db.ProdUserLog) error {
		var p OrderCreatedStartedPayload
		if err := decodeWithNumbers(row.Payload, &p); err != nil {
			return fmt.Errorf("parse order-created-started payload: %w", err)
		}
		if p.IDEquipment == 0 || p.IDOrder == "" {
			return fmt.Errorf("order-created-started payload missing fields: %+v", p)
		}
		idOrderNum, err := p.IDOrder.Int64()
		if err != nil {
			return fmt.Errorf("idOrder not numeric: %w", err)
		}
		qty, err := p.ProductionOrderQuantity.Float64()
		if err != nil {
			return fmt.Errorf("productionOrderQuantity not numeric: %w", err)
		}

		stagingEqID, err := t.Equipment(ctx, p.IDEquipment)
		if err != nil {
			return fmt.Errorf("translate equipment: %w", err)
		}
		stagingSiteID, err := t.Site(ctx, p.IDSite)
		if err != nil {
			return fmt.Errorf("translate site: %w", err)
		}
		stagingAreaID, err := t.Area(ctx, p.IDArea)
		if err != nil {
			return fmt.Errorf("translate area: %w", err)
		}

		var prodPOIDStr string
		// ts_creation: see comment in order_created.go (same prod-vs-staging column drift).
		found, err := prodDB.SelectOne(ctx,
			`SELECT id_production_order::text FROM production_orders
			  WHERE id_enterprise = $1 AND id_order = $2
			    AND ts_creation >= $3::timestamptz - interval '1 minute'
			  ORDER BY id_production_order DESC LIMIT 1`,
			[]any{cfg.ProdEnterpriseID, idOrderNum, row.TsEvent}, &prodPOIDStr)
		if err != nil {
			return fmt.Errorf("prod PO lookup: %w", err)
		}
		if !found {
			return fmt.Errorf("no prod production_order for id_enterprise=%d id_order=%d", cfg.ProdEnterpriseID, idOrderNum)
		}
		prodPOID, err := parseBigint(prodPOIDStr)
		if err != nil {
			return fmt.Errorf("parse prod PO id: %w", err)
		}

		if existing, hit, err := stagingDB.LookupMapping(ctx, "production_order", cfg.SourceName, prodPOID); err != nil {
			return fmt.Errorf("mapping lookup: %w", err)
		} else if hit {
			logger.Info("order-created-started already mapped, skipping",
				slog.Int64("sourceLogID", row.IDUserLogs),
				slog.Int64("prodPOID", prodPOID),
				slog.Int64("stagingPOID", existing))
			return nil
		}

		body := map[string]any{
			"idEnterprise":            cfg.StagingEnterpriseID,
			"idSite":                  stagingSiteID,
			"idArea":                  stagingAreaID,
			"idEquipment":             stagingEqID,
			"idOrder":                 idOrderNum,
			"productionOrderQuantity": qty,
			"timestamp":               p.Timestamp,
		}
		status, _, err := PostStaging(ctx, cfg, httpc, apiToken(), row,
			"/api/production-orders/create-and-start", body)
		if err != nil {
			return err
		}

		var stagingPOIDStr string
		found, err = stagingDB.SelectOne(ctx,
			`SELECT id_production_order::text FROM production_orders
			  WHERE id_enterprise = $1 AND id_order = $2
			  ORDER BY id_production_order DESC LIMIT 1`,
			[]any{cfg.StagingEnterpriseID, idOrderNum}, &stagingPOIDStr)
		if err != nil {
			return fmt.Errorf("staging PO lookup: %w", err)
		}
		if !found {
			return fmt.Errorf("create-and-start returned %d but staging PO with id_order=%d not found", status, idOrderNum)
		}
		stagingPOID, err := parseBigint(stagingPOIDStr)
		if err != nil {
			return fmt.Errorf("parse staging PO id: %w", err)
		}

		if err := db.InsertMapping(ctx, tx, db.MapInsert{
			EntityType:  "production_order",
			Source:      cfg.SourceName,
			ProdID:      prodPOID,
			StagingID:   stagingPOID,
			SourceLogID: row.IDUserLogs,
		}); err != nil {
			return fmt.Errorf("insert mapping: %w", err)
		}

		logger.Info("replayed order-created-started",
			slog.Int64("sourceLogID", row.IDUserLogs),
			slog.Int64("prodPOID", prodPOID),
			slog.Int64("stagingPOID", stagingPOID),
			slog.Int64("idOrder", idOrderNum),
			slog.Int("status", status))
		return nil
	}
}
